src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
   }
 };
 
+struct C_SetDedupChunks : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  uint64_t offset;
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0)
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    auto it = pg->manifest_ops.find(oid);
+    if (it == pg->manifest_ops.end()) {
+      // raced with cancel_manifest_ops
+      return;
+    }
+    // check if the previous op returns fail
+    if (it->second->results[0] < 0) {
+      return;
+    }
+
+    it->second->results[offset] = r;
+    if (last_peering_reset == pg->get_last_peering_reset() &&
+	r >= 0) {
+      if (it->second->num_chunks != it->second->results.size()) {
+	// there are on-going works
+	return;
+      }
+      pg->finish_set_dedup(oid, it->second->chunks, r, tid);
+    } else {
+      // if any failure occurs, put a mark on the results to recognize the failure
+      it->second->results[0] = r;
+      if (last_peering_reset != pg->get_last_peering_reset()) 
+	it->second->results[0] = -EINVAL;
```

Comment:
If num_chunks == results.size(), you need to call into finish_set_dedup as well, I think.  finish_set_dedup then neeeds to check results and return an error if any were unsuccessful.